src/NServiceBus.Core/Sagas/ActiveSagaInstance.cs

```diff
@@ -103,51 +114,35 @@ internal void ValidateChanges()
             {
                 ValidateAllCorrelationPropertiesHaveValues();
             }
-            else
-            {
-                ValidateCorrelationPropertiesNotModified();
-            }
+
+            ValidateCorrelationPropertiesNotModified();
         }
 
         void ValidateAllCorrelationPropertiesHaveValues()
         {
-            var properties = Instance.Entity.GetType().GetProperties();
-
-            foreach (var correlatedProperty in Metadata.CorrelationProperties)
+              foreach (var correlationProperty in correlationProperties)
             {
-                var propertyInfo = properties.Single(p => p.Name == correlatedProperty.Name);
-
-                var value = propertyInfo.GetValue(Instance.Entity);
-                var defaultValue = GetDefault(propertyInfo.PropertyType);
+                var defaultValue = GetDefault(correlationProperty.PropertyInfo.PropertyType);
 
-                if (value.Equals(defaultValue))
+                if (correlationProperty.Value.Equals(defaultValue))
                 {
                     throw new Exception(
-                        $@"We detected that the correlated property '{correlatedProperty.Name}' on saga '{Metadata.SagaType.Name}' does not have a value'. 
+                        $@"We detected that the correlated property '{correlationProperty.PropertyInfo.Name}' on saga '{Metadata.SagaType.Name}' does not have a value'. 
 All correlated properties must have a non null or empty value assigned to them when a new saga instance is created.");
                 }
-
-                correlatedPropertyValues[correlatedProperty.Name] = value;
             }
         }
 
         void ValidateCorrelationPropertiesNotModified()
         {
-            if (!correlatedPropertyValues.Any())
+            foreach (var correlationProperty in correlationProperties.Where(cp=>cp.HasExistingValue))
             {
-                return;
-            }
-
-            var properties = Instance.Entity.GetType().GetProperties();
-            foreach (var existingPropertyValue in correlatedPropertyValues)
-            {
-                var propertyInfo = properties.Single(p => p.Name == existingPropertyValue.Key);
-                var currentValue = propertyInfo.GetValue(Instance.Entity);
+                var currentValue = correlationProperty.PropertyInfo.GetValue(Instance.Entity);
 
-                if (!existingPropertyValue.Value.Equals(currentValue))
+                if (correlationProperty.Value.ToString() != currentValue.ToString())
```

Comment:
I think that's permissible. I shudder to think what a Saga would do if you tried to use some type that didn't have a well-understood, ToString()able value. I think the list should pretty clearly mimic those that have known ODBC equivalents (excluding binary blobs maybe) and those should all have ToString() implementations. Keeping it safe by default sounds good.